From 0b6eceb091415eef28486940c725507e7c680b1f Mon Sep 17 00:00:00 2001
From: Jose Ferreira <bazoocaze@gmail.com>
Date: Fri, 28 Aug 2015 09:14:38 -0300
Subject: [PATCH 4/4] Fix the cell standby mobile radio active issue

How to reproduce the bug:
-Enable mobile data
-Use some mobile data (radio active timer starts in battery stats)
-Disable mobile data (radio active timer keeps running in battery stats)

After some investigation on the issue I found that the BatteryStats
service was not receiving the radio power down notification. The
investigation lead me to the NetworkManagementService.java, where
I found that some code was discarding the radio power change
notifications after the first radio power on.

The problem is that in NetworkManagementServer.java the
notifyInterfaceClassActivity method is discarding the notifications
of radio power change to the BatteryStats (via noteMobileRadioPowerState)
if the fromRadio parameter is false after it's being true some time before.

Even after a detailed analysis, it was unclear what is the real role of
the fromRadio parameter. Changing it to true in the calling methods
solved the issue in the testing unit (Sony Xperia ZL). I started a thread
in XDA to mobilize some volunteers to help test this fix.

http://forum.xda-developers.com/xperia-zl/general/cm12-1-cell-standby-mobile-radio-active-t3188147

Change that introduced the code with the potential to trigger the bug:
 -Introduction of "boolean fromRadio" in
  NetworkManagementService.java / notifyInterfaceClassActivity()
 https://android.googlesource.com/platform/frameworks/base/+/2ffa11e4b71c545e34533ef827bdc1a07fbe8246

Change that triggered the bug:
 -Uncommenting the code that calls notifyInterfaceClassActivity
  with fromRadio=true
 https://android.googlesource.com/platform/frameworks/base/+/2d4b4ed681efcfb64c4d2b4d2a2d236aae1dc7ee

Change-Id: Icec9db692b0781dfb869858796ca8bb448db3f62
---
 services/core/java/com/android/server/NetworkManagementService.java | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/NetworkManagementService.java b/services/core/java/com/android/server/NetworkManagementService.java
index dded319..42b5f93 100644
--- a/services/core/java/com/android/server/NetworkManagementService.java
+++ b/services/core/java/com/android/server/NetworkManagementService.java
@@ -737,7 +737,7 @@ public class NetworkManagementService extends INetworkManagementService.Stub
                     boolean isActive = cooked[2].equals("active");
                     notifyInterfaceClassActivity(cooked[3] == null ? 0 : Integer.parseInt(cooked[3]),
                             isActive ? DataConnectionRealTimeInfo.DC_POWER_STATE_HIGH
-                            : DataConnectionRealTimeInfo.DC_POWER_STATE_LOW, timestampNanos, false);
+                            : DataConnectionRealTimeInfo.DC_POWER_STATE_LOW, timestampNanos, true);
                     return true;
                     // break;
             case NetdResponseCode.InterfaceAddressChange:
@@ -1512,7 +1512,7 @@ public class NetworkManagementService extends INetworkManagementService.Stub
                 @Override public void run() {
                     notifyInterfaceClassActivity(type,
                             DataConnectionRealTimeInfo.DC_POWER_STATE_HIGH,
-                            SystemClock.elapsedRealtimeNanos(), false);
+                            SystemClock.elapsedRealtimeNanos(), true);
                 }
             });
         }
@@ -1541,7 +1541,7 @@ public class NetworkManagementService extends INetworkManagementService.Stub
                 @Override public void run() {
                     notifyInterfaceClassActivity(params.type,
                             DataConnectionRealTimeInfo.DC_POWER_STATE_LOW,
-                            SystemClock.elapsedRealtimeNanos(), false);
+                            SystemClock.elapsedRealtimeNanos(), true);
                 }
             });
         }
-- 
1.9.1

